### PR TITLE
fix(automation): harden semantic rebase conflict resolution

### LINKED ISSUE #2779: Harden semantic rebase conflict resolution and pre-publish validation
## Problem

The automation loop can encounter a rebase conflict that is textually isolated but semantically spans multiple files. For example, two branches can add different ADRs using the same numeric identifier. The edit-only resolver is restricted to the conflicted file, so it may keep both rows in the index while leaving duplicate ADR numbers on disk.

During the #2754 run, the resolver produced a signed local rebased commit containing two ADR-0027 records. The host continuation checks passed because they validate conflict markers, scope, ancestry, signature, and DCO metadata, but they do not run repository-specific structural tests before publishing. The pre-push validation then rejected or timed out, while the event record retained only a generic `error`.

## Required behavior

- Detect when a conflict cannot be safely resolved by editing only the conflicted paths, including cross-file invariants such as ADR numbering and index consistency.
- Allow a safe conflict handoff to include all files required for a semantic resolution, or fail closed with an actionable manual-resolution reason.
- Run the repository's required focused validation on the resolved tree before attempting the lease-protected push.
- Classify publish, validation, and conflict-resolution failures separately; retain bounded stderr/stdout diagnostics in durable job state and pipeline events.
- Preserve the worktree and remote branch unchanged when validation or publication fails.
- Add regression tests for duplicate ADR numbering / index conflicts, invalid post-resolution trees, publish rejection, and diagnostic preservation.

## Acceptance criteria

- [ ] A conflict involving an ADR number collision is identified as semantic and is not resolved by blindly unioning README rows.
- [ ] The resolver can rename/update the affected ADR and index consistently, or returns an explicit actionable handoff without publishing.
- [ ] A resolved tree failing `tests/unit/docs/test_adr_records.py` is rejected before push.
- [ ] A failed push does not report as a generic exhausted conflict; it records whether the failure was validation, hook rejection, timeout, lease drift, or transport failure.
- [ ] The branch remains at its expected remote SHA after failed resolution/publication.
- [ ] Focused behavior-first tests cover the above cases.

Discovered while running the automation loop for #2754.

### PR DESCRIPTION
**Pull request by mvillmow**
## Summary

- Fail closed when edit-only rebase resolution leaves invalid ADR structure.
- Validate `tests/unit/docs/test_adr_records.py` against an immutable rebased snapshot before the lease push.
- Classify publication failures and retain bounded diagnostics for recovery and review.

## Root cause

The resolver used for #2754 could union README rows after an ADR-number collision. Existing host checks validated markers, scope, ancestry, signatures, and DCO, but not repository-level semantic invariants. Publication and hook failures were also classified too generically for reliable recovery.

## Behavior

Semantic conflicts now fail closed and preserve the worktree for manual resolution. Rebase publication records the failure phase, classification, bounded command output, and durable detail while preserving the existing success-event contract.

## Validation

- 429 affected tests passed.
- 112 coordinator/event regression tests passed.
- `tests/unit/docs/test_adr_records.py`: 4 passed.
- Ruff check and format passed.
- Mypy passed for the changed files.
- Full pre-push suite: 7,882 passed, 13 skipped, 7 deselected; 85.60% coverage.

Closes #2779

### CODE REVIEW

File: hephaestus/automation/pipeline/worker_pool.py [hunk 1]
```diff
@@ -88,6 +88,25 @@
 _TAIL = 4000  # chars of stdout/stderr retained in a JobResult
 _ERR_MAX = 500  # chars of error detail retained in a JobResult
 _GIT_LOCK_WAIT_POLL_S = 0.1
+_ADR_FILENAME_RE = re.compile(r"^(?P<number>[0-9]{4})-[a-z0-9-]+\.md$")
```
Comment (mvillmow) [inline hephaestus/automation/pipeline/worker_pool.py]: **major / required — keep repository-specific policy out of the shared worker**

`_validate_rebased_tree` is invoked for every rebase and the new module-level constants encode this repository’s ADR filename, section, README-index, and fixed `tests/unit/docs/test_adr_records.py` contract. `WorkerPool` is the shared multi-repository executor described in `docs/architecture.md` section 8, so another repository with a different valid `docs/adr` layout will now be rejected during rebase. Inject a host-trusted, repository-owned validator/argv (or narrow this policy to the owning repository) and add a regression proving a different ADR layout remains unaffected.

File: hephaestus/automation/git_utils.py [hunk 2]
```diff
@@ -406,10 +439,18 @@ def push_head_to_branch(
             return
         if not observed or observed[0] != expected_remote_sha:
             raise DetachedHeadPushRemoteHeadChangedError(
-                "Detached review push observed a different remote head"
+                "Detached review push observed a different remote head",
+                failure_kind="lease_drift",
             ) from exc
+        if isinstance(exc, subprocess.TimeoutExpired):
+            failure_kind = "timeout"
+        elif exc.returncode == 1:
```
Comment (mvillmow) [inline hephaestus/automation/git_utils.py:447]: **major / required — do not infer hook rejection from exit code 1**

A local pre-push hook rejection is not uniquely identified by `returncode == 1`; remote rejection and other Git failures can also return 1 while the remote ref remains unchanged. This can misclassify a remote/publication failure as `hook_rejected`, violating #2779’s separate failure classifications and producing misleading recovery state. Establish classification at an explicit hook boundary, or use an unknown/remote-rejected classification until hook evidence is available, and add a regression for unchanged-remote remote rejection.

File: hephaestus/automation/pipeline/coordinator_runtime.py [hunk 3]
```diff
@@ -763,15 +764,22 @@ def _record_resumable_recovery_worktrees(self, item: WorkItem) -> None:
...
             "ok": result.ok,
             "interrupted": result.interrupted,
             "error": CoordinatorRuntime._job_result_error_class(result),
             "duration_s": round(result.duration_s, 3),
         }
         if result.worker_id:
             fields["worker_id"] = result.worker_id
+        diagnostics = {}
+        if result.stdout_tail:
+            diagnostics["stdout_tail"] = result.stdout_tail[-_JOB_DIAGNOSTIC_MAX:]
```
Comment (mvillmow) [inline hephaestus/automation/pipeline/coordinator_runtime.py]: **major / required — sanitize diagnostics before durable persistence**

These raw stdout/stderr tails are copied into the durable JSONL event log, and the implementation stage also retains the same tails in item payload state. `_json_safe` only serializes values; it does not redact them. The existing completion contract explicitly describes records as output-free, while Git/pytest/tool output can contain source text, issue-controlled content, URLs, or credentials. A size bound is not a secrecy boundary. Apply a centralized redactor or allowlisted diagnostic schema before persistence, and add a regression proving secret-like stdout/stderr are removed or masked.

**Review by mvillmow — COMMENTED** [review]
## Athena PR review — NO-GO

Reviewed exact head `482bc62460f432683076aabd521e00aa5531f41e` against base `ac32bc21df961eb79fff267f679eeb8b720c19ad`. The PR is open and non-draft, and it links #2779. Exact-head evidence currently has `pr-policy` failing and `uv-lock-check` in progress, so merge readiness is not established.

### Artifact-level blocker — major / required

The repository policy requires a Conventional Commit PR title and Conventional Commit subjects for every authored branch commit. The current title is `Harden semantic rebase conflict resolution`; the two branch subjects are `Harden semantic rebase publication` and `Complete rebase validation diagnostics`. This is also the direct cause of the exact-head `pr-policy` failure. Rename the PR and rewrite the two branch commits with compliant subjects, preserving their signatures and DCO trailers, then rerun the checks.

### Scorecard

- Architecture/design: 8/30
- Issue/scope: 15/20
- Implementation: 11/18
- Testing/evidence: 11/15
- Security/safety: 4/10
- Integration/release: 2/7

Total: 51/100 (F).

The immutable validation boundary, lease-preserving publication behavior, typed failure state, and focused behavior tests are useful foundations. The required findings below must be resolved before this can be considered for GO.

**Comment by mvillmow** [comment]
### Host verification failed

[auto-msg] The fixed host-verification command failed before source review. The PR remains `state:implementation-no-go`.

**Reviewed head**

    b14788abe42386e947c4c5d67cb153f2cf3df6e3

**Verification command**

    uv run ruff check hephaestus/ tests/

**Failure classification**

    runner

**Failure**

    host_verification_quota_unavailable

**Comment by mvillmow** [comment]
### Host verification failed

[auto-msg] The fixed host-verification command failed before source review. The PR remains `state:implementation-no-go`.

**Reviewed head**

    b14788abe42386e947c4c5d67cb153f2cf3df6e3

**Verification command**

    uv run pytest -o addopts= tests/unit/automation/pipeline/test_worker_pool.py::TestAgentErrorHandling::test_codex_event_failure_is_explicit_agent_error -q --tb=short

**Affected path**

    tests/unit/automation/pipeline/test_worker_pool.py

**Failure classification**

    validation

**Failure**

    rc=1

**Standard output (tail)**

    F                                                                        [100%]
    =================================== FAILURES ===================================
    ___ TestAgentErrorHandling.test_codex_event_failure_is_explicit_agent_error ____
    tests/unit/automation/pipeline/test_worker_pool.py:1780: in test_codex_event_failure_is_explicit_agent_error
        assert result.error.startswith("agent_error: codex_nested_sandbox_unsupported")
    E   assert False
    E    +  where False = <built-in method startswith of str object at 0x103c17d70>('agent_error: codex_nested_sandbox_unsupported')
    E    +    where <built-in method startswith of str object at 0x103c17d70> = "PermissionError: [Errno 1] Operation not permitted: '/private/tmp'".startswith
    E    +      where "PermissionError: [Errno 1] Operation not permitted: '/private/tmp'" = JobResult(ok=False, value=None, stdout_tail='', stderr_tail='', error="PermissionError: [Errno 1] Operation not permit...vate/tmp'", interrupted=False, duration_s=0.0, worker_id='', session_id=None, session_binding=None, session_lost=False).error
    ------------------------------ Captured log call -------------------------------
    ERROR    hephaestus.automation.pipeline.worker_pool:worker_pool.py:232 Agent job raised, returning error result
    Traceback (most recent call last):
      File "/private/var/folders/67/4vl6jn7j4glfh89pgph32mxr0000gt/T/hephaestus-host-verification-zen0ql_4/source/hephaestus/automation/pipeline/worker_pool.py", line 1884, in _run_agent
        cwd = validate_job_workspace(job)
      File "/private/var/folders/67/4vl6jn7j4glfh89pgph32mxr0000gt/T/hephaestus-host-verification-zen0ql_4/source/hephaestus/automation/pipeline/jobs.py", line 94, in validate_job_workspace
        canonical = job.cwd.resolve(strict=True)
      File "/opt/homebrew/Caskroom/miniconda/base/lib/python3.13/pathlib/_local.py", line 670, in resolve
        return self.with_segments(os.path.realpath(self, strict=strict))
                                  ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^
      File "<frozen posixpath>", line 451, in realpath
    PermissionError: [Errno 1] Operation not permitted: '/private/tmp'
    =========================== short test summary info ============================
    FAILED tests/unit/automation/pipeline/test_worker_pool.py::TestAgentErrorHandling::test_codex_event_failure_is_explicit_agent_error
    1 failed in 0.78s
